consensus/core/src/core.rs

```diff
@@ -926,40 +939,42 @@ impl Core {
 
         let quorum_round = clock_round.saturating_sub(1);
 
-        let mut temp_excluded_ancestors = Vec::new();
-        let mut excluded_and_equivocating_ancestors = Vec::new();
+        let mut score_and_pending_excluded_ancestors = Vec::new();
+        let mut excluded_and_equivocating_ancestors = BTreeSet::new();
 
         // Propose only ancestors of higher rounds than what has already been proposed.
         // And always include own last proposed block first among ancestors.
         // Start by only including the high scoring ancestors. Low scoring ancestors
         // will be included in a second pass below.
         let included_ancestors = iter::once(self.last_proposed_block().clone())
             .chain(
-                ancestors
+                all_ancestors
                     .into_iter()
                     .filter(|(ancestor, _)| ancestor.author() != self.context.own_index)
                     .flat_map(|(ancestor, equivocating_ancestors)| {
-                        let ancestor_state = ancestor_state_map[ancestor.author()];
+                        if let Some(last_block_ref) =
+                            self.last_included_ancestors[ancestor.author()]
+                        {
+                            if last_block_ref.round >= ancestor.round() {
+                                return None;
+                            }
+                        }
 
                         // We will never include equivocating ancestors so add them immediately
                         excluded_and_equivocating_ancestors.extend(equivocating_ancestors);
 
+                        let ancestor_state = ancestor_state_map[ancestor.author()];
                         match ancestor_state {
                             AncestorState::Include => {
                                 trace!("Found ancestor {ancestor} with INCLUDE state for round {clock_round}");
                             }
                             AncestorState::Exclude(score) => {
                                 trace!("Added ancestor {ancestor} with EXCLUDE state with score {score} to temporary excluded ancestors for round {clock_round}");
-                                temp_excluded_ancestors.push((score, ancestor));
+                                score_and_pending_excluded_ancestors.push((score, ancestor));
                                 return None;
                             }
                         }
 
-                        if let Some(last_block_ref) =
```

Comment:
Why did you remove this filtering?

Reply:
I moved the filtering to earlier in the function.